src/radiation_system.hpp

```diff
@@ -917,189 +1022,244 @@ void RadSystem<problem_t>::AddSourceTerms(array_t &consVar, arrayconst_t &radEne
+		const double Erad0 = sum(Erad0Vec);
 
 		// load radiation energy source term
 		// plus advection source term (for well-balanced/SDC integrators)
-		const double Src = dt * ((chat * radEnergySource(i, j, k)) + advectionFluxes(i, j, k));
+		quokka::valarray<double, nGroups_> Src;
+		for (int g = 0; g < nGroups_; ++g) {
+			Src[g] = dt * (chat * radEnergySource(i, j, k, g) + advectionFluxes(i, j, k, g));
+		}
 
 		double Egas0 = NAN;
 		double Ekin0 = NAN;
 		double Egas_guess = NAN;
-		double Erad_guess = NAN;
 		double T_gas = NAN;
+		quokka::valarray<double, nGroups_> EradVec_guess{};
+		quokka::valarray<double, nGroups_> kappaPVec{};
+		quokka::valarray<double, nGroups_> kappaEVec{};
+		quokka::valarray<double, nGroups_> kappaFVec{};
 
-		if constexpr (gamma_ != 1.0) {
+		for (int g = 0; g < nGroups_; ++g) {
+			EradVec_guess[g] = NAN;
+		}
+
+		// if constexpr (gamma_ != 1.0) {
+		if (gamma_ != 1.0) {
 			Egas0 = ComputeEintFromEgas(rho, x1GasMom0, x2GasMom0, x3GasMom0, Egastot0);
 			Ekin0 = Egastot0 - Egas0;
 
-			AMREX_ASSERT(Src >= 0.0);
+			AMREX_ASSERT(min(Src) >= 0.0);
 			AMREX_ASSERT(Egas0 > 0.0);
-			AMREX_ASSERT(Erad0 > 0.0);
 
-			const double Etot0 = Egas0 + (c / chat) * (Erad0 + Src);
+			const double Etot0 = Egas0 + (c / chat) * (Erad0 + sum(Src));
 
 			// BEGIN NEWTON-RAPHSON LOOP
 			Egas_guess = Egas0;
-			Erad_guess = Erad0;
+			EradVec_guess = Erad0Vec;
 
 			const double a_rad = radiation_constant_;
 			double F_G = NAN;
-			double F_R = NAN;
-			double rhs = NAN;
-			double kappa = NAN;
 			double fourPiB = NAN;
-			double drhs_dEgas = NAN;
 			double dFG_dEgas = NAN;
-			double dFG_dErad = NAN;
-			double dFR_dEgas = NAN;
-			double dFR_dErad = NAN;
-			double eta = NAN;
-			double deltaErad = NAN;
 			double deltaEgas = NAN;
+			double Rtot = NAN;
+			double dRtot_dEgas = NAN;
+			quokka::valarray<double, nGroups_> Rvec;
+			quokka::valarray<double, nGroups_> dRtot_dErad;
+			quokka::valarray<double, nGroups_> dRvec_dEgas;
+			quokka::valarray<double, nGroups_> dFG_dErad;
+			quokka::valarray<double, nGroups_> dFR_dEgas;
+			quokka::valarray<double, nGroups_> dFR_i_dErad_i;
+			quokka::valarray<double, nGroups_> deltaErad;
+			quokka::valarray<double, nGroups_> F_R;
 
 			const double resid_tol = 1.0e-10; // 1.0e-15;
 			const int maxIter = 400;
 			int n = 0;
-			for (n = 0; n < maxIter; ++n) {
-
+			for (; n < maxIter; ++n) {
 				// compute material temperature
 				T_gas = quokka::EOS<problem_t>::ComputeTgasFromEint(rho, Egas_guess, massScalars);
 				AMREX_ASSERT(T_gas >= 0.);
 
 				// compute opacity, emissivity
-				kappa = RadSystem<problem_t>::ComputePlanckOpacity(rho, T_gas);
-				AMREX_ASSERT(kappa >= 0.);
 				fourPiB = chat * a_rad * std::pow(T_gas, 4);
+				auto B = fourPiB / (4.0 * pi);
+				kappaPVec = ComputePlanckOpacity(rho, T_gas);
+				kappaFVec = ComputeFluxMeanOpacity(rho, T_gas);
+				kappaEVec = kappaFVec; // TODO(CCH): later, define ComputeEnergyMeanOpacity
```

Comment:
Can you explain what you mean here?

Reply:
There are kappa_P, kappa_F, and kappa_E. Currently I'm assuming kappa_E = kappa_F. In the future we want to release this assumption.